listenbrainz/webserver/static/js/src/recommendations/RecommendationCard.tsx

```diff
@@ -1,36 +1,207 @@
 import * as React from "react";
+import {
+  faAngry,
+  faFrown,
+  faMeh,
+  faSmileBeam,
+  faGrinStars,
+} from "@fortawesome/free-solid-svg-icons";
+
+import {
+  faThumbsUp as faThumbsUpRegular,
+  faAngry as faAngryRegular,
+  faFrown as faFrownRegular,
+  faMeh as faMehRegular,
+  faSmileBeam as faSmileBeamRegular,
+  faGrinStars as faGrinStarsRegular,
+} from "@fortawesome/free-regular-svg-icons";
+
 import { get as _get } from "lodash";
-import { faEllipsisV } from "@fortawesome/free-solid-svg-icons";
 import { IconProp } from "@fortawesome/fontawesome-svg-core";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import RecommendationControl from "./RecommendationControl";
 
 import { getArtistLink, getTrackLink } from "../utils";
 import Card from "../components/Card";
+import APIService from "../APIService";
 
 export type RecommendationCardProps = {
+  apiUrl: string;
   recommendation: Recommendation;
   playRecommendation: (recommendation: Recommendation) => void;
+  currentFeedback: RecommendationFeedBack | null;
+  updateFeedback: (
+    recordingMbid: string,
+    rating: RecommendationFeedBack | null
+  ) => void;
   className?: string;
   isCurrentUser: Boolean;
   currentUser?: ListenBrainzUser;
+  newAlert: (
+    alertType: AlertType,
+    title: string,
+    message: string | JSX.Element
+  ) => void;
+};
+
+type RecommendationCardState = {
+  feedback: RecommendationFeedBack | null;
 };
 
 export default class RecommendationCard extends React.Component<
-  RecommendationCardProps
+  RecommendationCardProps,
+  RecommendationCardState
 > {
+  APIService: APIService;
   playRecommendation: (recommendation: Recommendation) => void;
 
   constructor(props: RecommendationCardProps) {
     super(props);
 
+    this.state = {
+      feedback: props.currentFeedback || null,
+      icon: faThumbsUpRegular,
+      text: "Like",
+    };
+
+    this.APIService = new APIService(
+      props.apiUrl || `${window.location.origin}/1`
+    );
+
     this.playRecommendation = props.playRecommendation.bind(
       this,
       props.recommendation
     );
   }
 
+  componentDidUpdate(
+    prevProps: RecommendationCardProps,
+    prevState: RecommendationCardState
+  ) {
+    const { currentFeedback } = this.props;
+    if (currentFeedback !== prevProps.currentFeedback) {
```

Comment:
@MonkeyDo I have changed all the bits as per the comments and everything works as expected.  But I noticed one thing, on refresh or fresh loading of the page, the feedback is rendered as expected, for example: if a user has hated a track it will be rendered as `faAngry Hate`, good enough as expected.

But

If I do `next` to move to the next page and do `previous` to move to the previous page the feedback changes from `faAngry Hate`  => `faThumbsUpRegular Like` which isn't expected. I checked the state and realised that the `icon` and `text` states change to the one initialized in constructor. 

```
this.state = {
      feedback: props.currentFeedback || null,
      icon: faThumbsUpRegular,
      text: "Like",
    };
```

The only solution I can think of is do `if (icon !== prevState.icon && text != prevState.text)` and if the condition evaluates to `True` i should update the icon and text accordingly.

Before moving forward, I wanted to ask if this solution makes sense to you and are there any drawbacks to this? I am not really sure about the downsides of `componentDidUpdate`

Posting a video on IRC for your reference.

Reply:
Two changes are needed to get the loading feedback to work with pagination:

1. Call `this.loadFeedback();` in `afterRecommendationsDisplay` (which is called after switching pages each time)
2. Change `getFeedback` to fetch feedback for the recordings in the *state* rather than in the props (i.e the current page the user sees instead of for all recs):
```
getFeedback = async () => {
    const { recommendations } = this.state;
    const { user } = this.props;
```

Reply:
@MonkeyDo I understand why we want to do the first step but I don't really understand the purpose of step 2, I mean I don't think step 2 is the reason for feedback not getting updated with `next` and `previous`. Are we doing it to make the rendering process faster (fetching feedback for 25 recs is faster than fetching feedback for 100 recs, though I don't think it really matters)?

Since if we fetch feedback for recording in the props (all recs), the current page recs feedback will also be updated since current page resc are a subset of all recs.

Reply:
You're absolutely right, in fact there should be no need to make a new request each time we change page.

Instead, there's something fishy happening in `componentDidUpdate` in the `RecommendationCard` component, and the icon is not updated. I thought it could be because `setState` is called twice in a row, but that didn't end up being the case. I'm not entirely sure why, but `currentFeedback` is always equal to `prevProps.currentFeedback` and the icon+text aren't updated.

Here's what I've tested and propose:

1. Get rid of `componentDidUpdate` entirely
2. We can get rid of all the items in the state. We don't need to copy the props into state, instead we can use the props directly in the render function
3. Place this snippet in the render function:
```
    const { currentFeedback, recommendation, className } = this.props;
    let icon: any;
    let text: string;
    switch (currentFeedback) {
      case "hate":
        icon = faAngry;
        text = "hate";
        break;
      case "dislike":
        icon = faFrown;
        text = "dislike";
        break;
      case "bad_recommendation":
        icon = faMeh;
        text = "bad";
        break;
      case "like":
        icon = faSmileBeam;
        text = "like";
        break;
      case "love":
        icon = faGrinStars;
        text = "love";
        break;
      default:
        icon = faThumbsUpRegular;
        text = "Like";
        break;
    }
```

This switch statement is equivalent to if/else statements and sets icon and text according to the currentFeedback prop (the prop is always up-to-date in the render function).
Note that I'm not using the state but rather the props, and you'll need to replace `feedback` with `currentFeedback` where you instantiate the RecommendationControl components.

With this the cards are updated when the props change and all looks fine :)

Reply:
> currentFeedback is always equal to prevProps.currentFeedback

isn't this something to worry about? Ideally they should not be equal. Can the code break in future if we don't compare `currentFeedback` with `prevProps.currentFeedback`? isn't it an alarm if the state is not changing?

Reply:
Should we investigate where the value is coming from in `currentFeedback`?

Reply:
>Should we investigate where the value is coming from in currentFeedback?

I found the root issue: `componentDidUpdate` is not called when the component is initially instantiated, but only on subsequent re-renders

From the docs: https://reactjs.org/docs/react-component.html#componentdidupdate
>componentDidUpdate() is invoked immediately after updating occurs. This method is not called for the initial render.

I think this point out to the fact that we don't need `componentDidUpdate` here, and using such mechanisms to copy the props into state is a bug-prone anti-pattern (for this component).

There's a good explanation why [here](https://daveceddia.com/where-initialize-state-react/)
>Initializing State from Props
In most cases, this is an antipattern. Don’t “copy props into state.” It creates a second source of truth for your data, which usually leads to bugs. One source of truth is best.
Components will already re-render when their props change, so there’s no need to duplicate the props as state and then try to keep it up to date.

>Think of it this way: it’s fine if the state needs a starting value which the component will then control. Ask yourself: does this component “own” the data? Does it just need a jump-start from a prop? Those are good reasons to initialize state from a prop.

For example, in the `Recommendations` component you do use `this.props.recommendations` as a source to set a slice of it in the state (https://github.com/metabrainz/listenbrainz-server/blob/rec-component-feedback/listenbrainz/webserver/static/js/src/recommendations/Recommendations.tsx#L47).
It makes sense there to initialize the state from props, as we don't expect the props to change.